File: observer_java/src/main/java/com/yb/observer_java/observer/ObserverManager.java
package com.yb.observer_java.observer;

import com.yb.observer_java.listener.ObserverListener;
import com.yb.observer_java.listener.SubjectListener;

import java.util.ArrayList;
import java.util.List;

/**
 * 类说明：观察者模式管理类
 *
 * @author 裕博
 */
public class ObserverManager implements SubjectListener {

    private static ObserverManager observerManager;
    /**
     * 观察者集合
     */
    private List<ObserverListener> list = new ArrayList<>();

    /**
     * 单例
     */
    public static ObserverManager getInstance() {
        if (null == observerManager) {
            synchronized (ObserverManager.class) {
                if (null == observerManager) {
                    observerManager = new ObserverManager();
                }
            }
        }
        return observerManager;
    }

    @Override
    public void add(ObserverListener observerListener) {
        // 加入队列
        list.add(observerListener);
    }

    @Override
    public void notifyObserver(String count) {
        // 通知观察者刷新数据
        for (ObserverListener observerListener : list) {
            observerListener.observerUpData(count);
        }
    }

    @Override
    public void remove(ObserverListener observerListener) {
        // 从监听队列删除
        list.remove(observerListener);
    }
}

File: observer_kotlin/src/main/java/com/yb/observer_kotlin/observer/ObserverManager.kt
package com.yb.observer_kotlin.observer

import com.yb.observer_kotlin.listener.ObserverListener
import com.yb.observer_kotlin.listener.SubjectListener
import java.util.ArrayList

/**
 * 类说明：观察者模式管理类
 *
 * @author 裕博
 */
class ObserverManager : SubjectListener {
    /**
     * 观察者集合
     */
    private val list = ArrayList<ObserverListener>()

    override fun add(observerListener: ObserverListener) {
        // 加入队列
        list.add(observerListener)
    }

    override fun notifyObserver(count: String) {
        // 通知观察者刷新数据
        for (observerListener in list) {
            observerListener.observerUpData(count)
        }
    }

    override fun remove(observerListener: ObserverListener) {
        // 从监听队列删除
        list.remove(observerListener)
    }

    companion object {
        val instance: ObserverManager by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
            ObserverManager()
        }
    }
}
